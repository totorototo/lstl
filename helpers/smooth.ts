type Getter<T> = (value: T) => number;
type Setter<T> = (value: T, smoothedValue: number) => number;

export function smooth<T>(
  arr: T[],
  windowSize: number,
  getter: Getter<T> = (value) => Number(value),
  setter?: Setter<T>,
): number[] {
  const get: Getter<T> = getter;
  const result: number[] = [];

  for (let i = 0; i < arr.length; i += 1) {
    const leftOffset = i - windowSize;
    const from = leftOffset >= 0 ? leftOffset : 0;
    const to = i + windowSize + 1;

    let count = 0;
    let sum = 0;
    for (let j = from; j < to && j < arr.length; j += 1) {
      sum += get(arr[j]);
      count += 1;
    }

    result[i] = setter ? setter(arr[i], sum / count) : sum / count;
  }

  return result;
}
