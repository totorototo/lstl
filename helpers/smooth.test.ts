import { describe, expect, it } from "vitest";
import { smooth } from "./smooth";

describe("smooth test suite", () => {
  it("should smooth the array", () => {
    // arrange
    const sample = [1, 2, 3, 4, 5, 4, 5, 6, 7, 8, 9];

    // act
    const result = smooth(sample, 3);

    // assert
    expect(result.length).toEqual(sample.length);
    expect(result).toMatchSnapshot();
  });
});
