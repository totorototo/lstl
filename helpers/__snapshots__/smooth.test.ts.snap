// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`smmoth test suite > should smooth the array 1`] = `
[
  2.5,
  3,
  3.1666666666666665,
  3.4285714285714284,
  4.142857142857143,
  4.857142857142857,
  5.571428571428571,
  6.285714285714286,
  6.5,
  7,
  7.5,
]
`;
